### PR TITLE
Refactor export map builder

### PR DESCRIPTION
**Pull request by soryy708**
[exportMapBuilder.js](https://github.com/import-js/eslint-plugin-import/blob/2de78c1eac5064858cc366e211913c4f9e43919b/src/exportMapBuilder.js) contains a huge god-object with 651 lines of code, with tightly coupled procedures that are hard to follow.

Lets extract procedures and classes, and put them in well named files.
This separation will also help with encapsulation and decoupling.

It's not perfect yet, but a reduction from 651 lines to 207 lines is a 68% decrease.

### CODE REVIEW

File: src/exportMap/patternCapture.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+/**
+ * Traverse a pattern/identifier node, calling 'callback'
+ * for each leaf identifier.
+ * @param  {node}   pattern
+ * @param  {Function} callback
+ * @return {void}
+ */
+export function recursivePatternCapture(pattern, callback) {
```
Comment (ljharb) [inline src/exportMap/patternCapture.js:8]:
```suggestion
export default function recursivePatternCapture(pattern, callback) {
```

File: src/exportMap/captureDependency.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
+export function captureDependency(
```
Comment (ljharb) [inline src/exportMap/captureDependency.js:1]:
```suggestion
export default function captureDependency(
```

File: src/exportMap/childContext.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+let parserOptionsHash = '';
+let prevParserOptions = '';
+let settingsHash = '';
+let prevSettings = '';
+
+/**
+ * don't hold full context object in memory, just grab what we need.
+ * also calculate a cacheKey, where parts of the cacheKey hash are memoized
+ */
+export function childContext(path, context) {
```
Comment (ljharb) [inline src/exportMap/childContext.js:12]:
```suggestion
export default function childContext(path, context) {
```

File: src/exportMap/namespace.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import { childContext } from './childContext';
+import { RemotePath } from './remotePath';
+
+export class Namespace {
```
Comment (ljharb) [inline src/exportMap/namespace.js:4]:
```suggestion
export default class Namespace {
```

File: src/exportMap/specifier.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
+export function processSpecifier(specifier, astNode, exportMap, namespace) {
```
Comment (ljharb) [inline src/exportMap/specifier.js:1]:
```suggestion
export default function processSpecifier(specifier, astNode, exportMap, namespace) {
```

File: src/exportMap/visitor.js [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+  if (SourceCode.length > 1) {
+    // ESLint 3
+    return new SourceCode(text, ast);
+  } else {
+    // ESLint 4, 5
+    return new SourceCode({ text, ast });
+  }
+}
+
+export class ImportExportVisitorBuilder {
```
Comment (ljharb) [inline src/exportMap/visitor.js:23]:
```suggestion
export default class ImportExportVisitorBuilder {
```

**Review by ljharb — COMMENTED** [review]
Thanks, this is great!

It seems like this refactor is exposing a lot of previously uncovered lines. It's fine to land this without covering them, but it would be even better if this PR (or another one, landed first) could include additional tests so we can ensure the uncovered lines are all covered

File: src/exportMap/patternCapture.js (hunk 1 above)
Comment (soryy708) [reply src/exportMap/patternCapture.js:8]: Sounds like we want to dog-food `prefer-default-export`

- https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/prefer-default-export.md

File: src/exportMap/captureDependency.js (hunk 2 above)
Comment (soryy708) [reply src/exportMap/captureDependency.js:1]: Why would `captureDependency` be the default and not, say, `captureDependencyWithSpecifiers` that is exported from the same file?

File: src/exportMap/patternCapture.js (hunk 1 above)
Comment (soryy708) [reply src/exportMap/patternCapture.js]: - https://github.com/import-js/eslint-plugin-import/pull/2991/commits/b4b1740153f293ef49982ae286fa1838a711c14d

File: src/exportMap/childContext.js (hunk 3 above)
Comment (soryy708) [reply src/exportMap/childContext.js]: - https://github.com/import-js/eslint-plugin-import/pull/2991/commits/b0294f8af19d5067e500b219c751dfe239b7d968

File: src/exportMap/namespace.js (hunk 4 above)
Comment (soryy708) [reply src/exportMap/namespace.js]: - https://github.com/import-js/eslint-plugin-import/pull/2991/commits/fe2c2801fdc39f9b5db70aa10d536d32c3dfa0d8

File: src/exportMap/specifier.js (hunk 5 above)
Comment (soryy708) [reply src/exportMap/specifier.js]: - https://github.com/import-js/eslint-plugin-import/pull/2991/commits/8df02eb508dd8d8eaea6cfdd89040bf663783993

File: src/exportMap/visitor.js (hunk 6 above)
Comment (soryy708) [reply src/exportMap/visitor.js]: - https://github.com/import-js/eslint-plugin-import/pull/2991/commits/c986ba3a6eceb813ddea1e8440390c7fbd3c771f

**Comment by soryy708** [comment]
> this refactor is exposing a lot of previously uncovered lines. It's fine to land this without covering them

I reviewed the lines that aren't covered.
Lets land this without covering them for now.
I can't think of good & quick tests to write for them at this time.

**Review by ljharb — APPROVED** [review]

File: src/exportMap/captureDependency.js (hunk 2 above)
Comment (ljharb) [reply src/exportMap/captureDependency.js:1]: because the filename is "captureDependency". a default export is what a module _is_, a named export is something a module _has_, and most modules should _be_ something.

**Comment by soryy708** [comment]
Has been merged via

- https://github.com/import-js/eslint-plugin-import/commit/8587c85a60ccb3839cb2eabe50cf098a4a2c03c2
